cpp/src/parquet/variant.h

```diff
@@ -0,0 +1,297 @@
+  Date = 11,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=true, MICROS)
+  Timestamp = 12,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=false, MICROS)
+  TimestampNtz = 13,
+  /// Equivalent Parquet Type: FLOAT
+  Float = 14,
+  /// Equivalent Parquet Type: BINARY
+  Binary = 15,
+  /// Equivalent Parquet Type: STRING
+  String = 16,
+  /// Equivalent Parquet Type: TIME(isAdjustedToUTC=false, MICROS)
+  TimeNtz = 17,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=true, NANOS)
+  TimestampTz = 18,  // Assuming TZ stands for TimeZone, and follows the document's
+                     // 'timestamp with time zone'
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=false, NANOS)
+  TimestampNtzNanos = 19,  // Differentiating from TimestampNtz (MICROS)
+  /// Equivalent Parquet Type: UUID
+  Uuid = 20
+};
+
+PARQUET_EXPORT std::string_view VariantPrimitiveTypeToString(VariantPrimitiveType type);
+
+/// VariantType is from basic type and primitive type.
+enum class VariantType {
+  Object,
+  Array,
+  Null,
+  Boolean,
+  Int8,
+  Int16,
+  Int32,
+  Int64,
+  String,
+  Double,
+  Decimal4,
+  Decimal8,
+  Decimal16,
+  Date,
+  TimestampTz,
+  TimestampNtz,
+  Float,
+  Binary,
+  Time,
+  TimestampNanosTz,
+  TimestampNanosNtz,
+  Uuid
+};
+
+PARQUET_EXPORT std::string_view VariantTypeToString(VariantType type);
+
+class PARQUET_EXPORT VariantMetadata {
+ public:
+  explicit VariantMetadata(std::string_view metadata);
+  /// \brief Get the variant metadata version. Currently, always 1.
+  uint8_t version() const;
+  /// \brief Get the metadata key for a given variant field id.
+  /// \throw ParquetException if the variant_id is out of range(larger than
+  ///        dictionary_size).
+  std::string_view GetMetadataKey(uint32_t variant_id) const;
+  /// \brief Get the metadata id for a given key.
+  /// From the discussion in ML:
+  /// https://lists.apache.org/thread/b68tjmrjmy64mbv9dknpmqs28vnzjj96 if
+  /// !sorted_and_unique(), the metadata key is not guaranteed to be unique, so we use a
+  /// vector to store all the metadata ids.
+  ::arrow::internal::SmallVector<uint32_t, 1> GetMetadataId(std::string_view key) const;
+
+  bool sorted_and_unique() const;
+  uint8_t offset_size() const;
+  uint32_t dictionary_size() const;
+
+ private:
+  static uint32_t loadDictionarySize(std::string_view metadata, uint8_t offset_size);
+
+ private:
+  static constexpr uint8_t kVersionMask = 0b1111;
+  static constexpr uint8_t kSortedStringMask = 0b10000;
+  static constexpr size_t kHeaderSizeBytes = 1;
+  static constexpr size_t kMinimalOffsetSizeBytes = 1;
+  static constexpr size_t kMaximumOffsetSizeBytes = 4;
+  // mask is applied after shift, it's like 0b11000000 before shift.
+  static constexpr uint8_t kOffsetSizeMask = 0b11;
+  static constexpr uint8_t kOffsetSizeBitShift = 6;
+  static constexpr uint8_t kSupportedVersion = 1;
+
+ private:
+  std::string_view metadata_;
+  uint32_t dictionary_size_{0};
```

Comment:
I wonder if it's worth to change this to below case, which could make this 24B -> 16B

```
const uint8_t* metadata_ptr_;
const uint32_t metadata_size_;
uint32_t dictionary_size_
```